src/hierarchies/base_hierarchy.h

```diff
@@ -22,326 +23,404 @@
+
   //! Returns an independent, data-less copy of this object
   std::shared_ptr<AbstractHierarchy> clone() const override {
+    // Create copy of the hierarchy
     auto out = std::make_shared<Derived>(static_cast<Derived const &>(*this));
-    out->clear_data();
-    out->clear_summary_statistics();
+    // Cloning each component class
+    out->set_likelihood(std::static_pointer_cast<Likelihood>(like->clone()));
+    out->set_prior(std::static_pointer_cast<PriorModel>(prior->clone()));
     return out;
-  }
+  };
 
-  //! Returns an independent, data-less copy of this object
+  //! Returns an independent, data-less deep copy of this object
   std::shared_ptr<AbstractHierarchy> deep_clone() const override {
+    // Create copy of the hierarchy
     auto out = std::make_shared<Derived>(static_cast<Derived const &>(*this));
+    // Simple clone for Likelihood is enough
+    out->set_likelihood(std::static_pointer_cast<Likelihood>(like->clone()));
+    // Deep clone required for PriorModel
+    out->set_prior(std::static_pointer_cast<PriorModel>(prior->deep_clone()));
+    return out;
+  }
 
-    out->clear_data();
-    out->clear_summary_statistics();
+  //! Returns an independent, data-less copy of this object
+  // std::shared_ptr<AbstractHierarchy> deep_clone() const override {
+  //   auto out = std::make_shared<Derived>(static_cast<Derived const
+  //   &>(*this));
+
+  //   out->clear_data();
+  //   out->clear_summary_statistics();
+
+  //   out->create_empty_prior();
+  //   std::shared_ptr<google::protobuf::Message> new_prior(prior->New());
+  //   new_prior->CopyFrom(*prior.get());
+  //   out->get_mutable_prior()->CopyFrom(*new_prior.get());
+
+  //   out->create_empty_hypers();
+  //   auto curr_hypers_proto = get_hypers_proto();
+  //   out->set_hypers_from_proto(*curr_hypers_proto.get());
+  //   out->initialize();
+  //   return out;
+  // }
+
+  //! Public wrapper for `like_lpdf()` methods
+  double get_like_lpdf(const Eigen::RowVectorXd &datum,
+                       const Eigen::RowVectorXd &covariate =
+                           Eigen::RowVectorXd(0)) const override {
+    return like->lpdf(datum, covariate);
+  }
 
-    out->create_empty_prior();
-    std::shared_ptr<google::protobuf::Message> new_prior(prior->New());
-    new_prior->CopyFrom(*prior.get());
-    out->get_mutable_prior()->CopyFrom(*new_prior.get());
+  //! Evaluates the log-likelihood of data in a grid of points
+  //! @param data        Grid of points (by row) which are to be evaluated
+  //! @param covariates  (Optional) covariate vectors associated to data
+  //! @return            The evaluation of the lpdf
+  Eigen::VectorXd like_lpdf_grid(const Eigen::MatrixXd &data,
+                                 const Eigen::MatrixXd &covariates =
+                                     Eigen::MatrixXd(0, 0)) const override {
+    return like->lpdf_grid(data, covariates);
+  };
+
+  // ADD EXCEPTION HANDLING
```

Comment:
It is a general doubt rather than a specific TODO, which would eventually require some changes in this "``lpdf`` - like" code patterns. Indeed, using the ``Eigen::RowVectorXd(0)`` as default for the ``covariate`` argument, may produce an unwanted behaviour in case you call, e.g.,  ``Likelihood::lpdf(datum)`` from a dependent hierarchy. You would expect an exception but it would produce an error since it would still call ``lpdf(datum, covariate)``, as ``is_dependent() = true`` and ``covariates = Eigen::RowVectorXd(0)``.

I noticed this issue while testing the univariate linear regression hierarchy, and I set a temporary fix to manage exception correctly in ``src/hierarchies/likelihoods/abstract_likelihood.h:25``. I do not know if this is ideal, but I add those TODO as reminders to share with you.